packages/core/src/controllers/ConvertController.ts

```diff
@@ -0,0 +1,687 @@
+import { RouterController } from './RouterController.js'
+import { NumberUtil } from '@web3modal/common'
+
+export const INITIAL_GAS_LIMIT = 150000
+
+// -- Types --------------------------------------------- //
+type TransactionParams = {
+  data: `0x${string}`
+  to: `0x${string}`
+  gas: bigint
+  gasPrice: bigint
+  value: bigint
+  toAmount: string
+}
+
+class TransactionError extends Error {
+  shortMessage?: string
+
+  constructor(message?: string, shortMessage?: string) {
+    super(message)
+    this.name = 'TransactionError'
+    this.shortMessage = shortMessage
+  }
+}
+
+export interface ConvertControllerState {
+  // Loading states
+  initialized: boolean
+  loadingPrices: boolean
+  loading?: boolean
+
+  // Approval & Convert transaction states
+  approvalTransaction: TransactionParams | undefined
+  convertTransaction: TransactionParams | undefined
+  transactionLoading?: boolean
+  transactionError?: string
+
+  // Input values
+  sourceToken?: TokenInfo
+  sourceTokenAmount: string
+  sourceTokenPriceInUSD: number
+  toToken?: TokenInfo
+  toTokenAmount: string
+  toTokenPriceInUSD: number
+  networkPrice: string
+  networkBalanceInUSD: string
+  inputError: string | undefined
+
+  // Request values
+  slippage: string
+
+  // Tokens
+  tokens?: Record<string, TokenInfo>
+  suggestedTokens?: Record<string, TokenInfo>
+  popularTokens?: Record<string, TokenInfo>
+  foundTokens?: TokenInfo[]
+  myTokensWithBalance?: Record<string, TokenInfoWithBalance>
+  tokensPriceMap: Record<string, string>
+
+  // Calculations
+  gasFee: bigint
+  gasPriceInUSD?: number
+  priceImpact: number | undefined
+  maxSlippage: number | undefined
+}
+
+export interface TokenInfo {
+  address: `0x${string}`
+  symbol: string
+  name: string
+  decimals: number
+  logoURI: string
+  domainVersion?: string
+  eip2612?: boolean
+  isFoT?: boolean
+  tags?: string[]
+}
+
+export interface TokenInfoWithPrice extends TokenInfo {
+  price: string
+}
+
+export interface TokenInfoWithBalance extends TokenInfo {
+  balance: string
+  price: string
+}
+
+type StateKey = keyof ConvertControllerState
+
+// -- State --------------------------------------------- //
+const state = proxy<ConvertControllerState>({
+  // Loading states
+  initialized: false,
+  loading: false,
+  loadingPrices: false,
+
+  // Approval & Convert transaction states
+  approvalTransaction: undefined,
+  convertTransaction: undefined,
+  transactionError: undefined,
+  transactionLoading: false,
+
+  // Input values
+  sourceToken: undefined,
+  sourceTokenAmount: '',
+  sourceTokenPriceInUSD: 0,
+  toToken: undefined,
+  toTokenAmount: '',
+  toTokenPriceInUSD: 0,
+  networkPrice: '0',
+  networkBalanceInUSD: '0',
+  inputError: undefined,
+
+  // Request values
+  slippage: ConstantsUtil.CONVERT_SLIPPAGE_TOLERANCE,
+
+  // Tokens
+  tokens: undefined,
+  popularTokens: undefined,
+  suggestedTokens: undefined,
+  foundTokens: undefined,
+  myTokensWithBalance: undefined,
```

Comment:
I think yeah. When we don't define undefined values in the initial state, I feel like it might be confusing, it gives more confidence when we see all of the states in a place. I do think we should keep them.